patroni/postgresql/__init__.py

```diff
@@ -711,9 +718,42 @@ def _wait_promote(self, wait_seconds):
             if data.get('Database cluster state') == 'in production':
                 return True
 
+    def _call_pre_promote(self, cmd):
+        if cmd:
+            try:
+                ret = self.cancellable.call(shlex.split(cmd))
+            except OSError as err:
+                logger.error('pre_promote script %s failed: %s', cmd, err)
+                return False
+
+            if ret != 0:
+                logger.error('pre_promote script %s returned non-zero code %d', cmd, ret)
+                return False
+
+        return True
+
+    def _pre_promote(self):
+        """
+        Runs a fencing script after the leader lock is acquired but before the replica is promoted.
+        If the script exits with a non-zero code, promotion does not happen and the leader key is removed from DCS.
+        """
+
+        self.pre_promote_task = CriticalTask()
+        try:
+            self.pre_promote_task.complete(self._call_pre_promote(self.pre_promote_script))
+        except Exception:
+            logger.exception('pre_promote')
+            self.pre_promote_task.complete(False)
+
+        return self.pre_promote_task.result
+
     def promote(self, wait_seconds, on_success=None, access_is_restricted=False):
         if self.role == 'master':
             return True
+
+        if self.pre_promote_script and not self._pre_promote():
```

Comment:
Here we've got a race condition with https://github.com/zalando/patroni/pull/1099/files#diff-57c53e988b775ee2902488fb33c1f433R1131.
In order to solve it we probably need to use the approch like in the [Postgresql.start()](https://github.com/zalando/patroni/blob/master/patroni/postgresql/__init__.py#L421-L432) method.

Reply:
is it the race condition between cancelling and starting `pre_promote` ? if so, isn't it already handled by the combination of [self._cancellable.cancel()](https://github.com/zalando/patroni/blob/master/patroni/async_executor.py#L127) and [raise PostgresException('cancelled')](https://github.com/zalando/patroni/blob/master/patroni/postgresql/cancellable.py#L92) ?

Reply:
> is it the race condition between canceling and starting pre_promote?

Yes

> isn't it already handled by the combination of self._cancellable.cancel() and raise PostgresException('cancelled') ?

No, it is not. Don't forget about GIL. For example the cancel call might happen after the script was successfully finished but before `self.pg_ctl('promote', '-W')`